File: 0436-find-right-interval/Solution.go
package Leetcode

import "sort"

func findRightInterval(intervals [][]int) []int {
	if len(intervals) == 0 {
		return []int{}
	}
	starts := make([][]int, len(intervals))
	for i := 0; i < len(intervals); i++ {
		starts[i] = []int{intervals[i][0], i}
	}
	sort.Slice(starts, func(a, b int) bool {
		return starts[a][0] < starts[b][0]
	})

	output := make([]int, len(intervals))

	for index := range intervals {
		end := intervals[index][1]
		left := 0
		right := len(intervals) - 1

		for left <= right {
			mid := left + (right-left)/2
			if starts[mid][0] >= end {
				right = mid - 1
			} else {
				left = mid + 1
			}
		}
		output[index] = -1
		if left != len(intervals) {
			output[index] = starts[left][1]
		}
	}
	return output
}

File: 0436-find-right-interval/Solution.cs
public class Solution
{
    public int[] FindRightInterval(int[][] intervals)
    {
        if (intervals.Length == 0) return new int[] { -1 };
        
        int[][] starts = new int[intervals.Length][];
        for (int i = 0; i < intervals.Length; i++)
        {
            starts[i] = new int[] { intervals[i][0], i };
        }
        
        Array.Sort(starts, (a, b) => a[0].CompareTo(b[0]));

        var output = new int[intervals.Length];
        
        for (var index = 0; index < intervals.Length; index++)
        {
            int end = intervals[index][1];
            int left = 0;
            int right = intervals.Length - 1;

            while (left <= right)
            {
                int mid = left + (right - left) / 2;
                if (starts[mid][0] < end)
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }

            output[index] = left == intervals.Length ? -1 : starts[left][1];
        }

        return output;
    }
}